src/prefect/engine.py

```diff
@@ -750,6 +750,11 @@ async def begin_task_map(
 ) -> List[Union[PrefectFuture, Awaitable[PrefectFuture]]]:
     """Async entrypoint for task mapping"""
 
+    # Preserve task dependencies to futures passed as parameters,
+    # as when they are resolved the relationship will be lost.
+    wait_for = [] if wait_for is None else wait_for
+    wait_for.extend([x for x in parameters.values() if isinstance(x, PrefectFuture)])
```

Comment:
Thanks a bunch for digging into this issue! It's really helpful. Unfortunately there a couple issues with this implementation:
- This will drop the parameter _key_ that the future is passed under. We need to preserve that data but it looks like we'll need to do some refactoring to pass it downstream.
- This will not find futures nested inside parameters. We use a very magical `visit_collections` utility for this.

We can take this on, but if you're interested in looking further the typical pattern for dependency collection is at https://github.com/PrefectHQ/prefect/blob/a48eb57ded87300a165090517a829afcf54e5c4b/src/prefect/engine.py#L959

Reply:
Thanks for taking the time to take a look!

> This will drop the parameter key that the future is passed under. We need to preserve that data but it looks like we'll need to do some refactoring to pass it downstream.

I agree that dropping the key is a bad idea, however, I'm not too clear on where I should be persisting that data to. 

> This will not find futures nested inside parameters. We use a very magical `visit_collections` utility for this.

Ah cheers, I realised that this would not work for chained maps so have a version working for those, but this should be what I need to generalise it even further.

Very happy to keep working away at this, but if it's going to be too much bandwidth for you guys to review and you'd rather do it in-house just let me know!

Reply:
If you're interested you're welcome to take it further. I think you'll need to handle a few complex cases though and then there will be a little bit of engine refactoring to pass the relevant data downstream.

I think the most helpful next step is to write test cases (that will fail for now) covering the each way that task relationships can get created by futures with map. Here are a few example cases I'm thinking about:

```python

@task
def foo(x):
   return x

@task
def bar(x, y):
   return x + y

@task
def many_foo():
    return [1,2,3]
 
@flow
def my_flow():
   x = foo.submit(1)
   bar.map([x, 2], y=1)   # Relationship between mapped child bar-0 and foo-0 but not mapped child bar-1
   bar.map([1, 2, 3], y=x)   # Relationship between all mapped children of bar and foo-0
   x2 = foo.submit(2)
   bar.map([x, x2], y=0)  # Relationships between foo-0/mapped child bar-0 and foo-1/mapped child bar-1
   xs = many_foo.submit()
   bar.map(xs, y=0) # Relationships between all mapped children of bar and many_foo
```

I'm not sure what the best way to handle this will be yet! We're off until Tuesday, but we can try to provide some  more specific guidance on the best way to pass the data around then. If it looks like it'll be painful or really hard we can take a more active role.

Reply:
Excellent! I have added those test cases plus one for more deeply nested maps. I will push a `wait_for` based solution to handle these cases, which can hopefully be useful during the refactor.